thoughts-of-a-rustacean-learning-go.md

```diff
@@ -0,0 +1,244 @@
+(for debug builds, obviously), but I'd certainly like it to be there some day.
+
+Go has really great <s>green threads</s> goroutines. They're rather efficient
+(I can spawn a thousand and it schedules them nicely), and easy to use.
+
+Go has really good in built support and tooling for tests (Rust does too).
+I enjoyed writing tests in Go quite a bit due to this.
+
+## What I didn't like
+
+Sadly, there are a lot of things here, but bear in mind what I mentioned above
+about me being new to Go and not yet familiar with the "Go way" of doing things.
+
+### No enums
+
+Rust has enums, which are basically tagged unions. Different variants can contain different types
+of data, so we can have, for example:
+
+```rust
+enum Shape {
+    Rectangle(Point, Point),
+    Circle(Point, u8),
+    Triangle(Point, Point, Point)
+}
+```
+
+and when matching/destructuring, you get type-safe access to the contents of the variant.
+
+This is extremely useful for sending typed messages across channels. For example, in Servo we use such
+[an enum for sending details about the progress of a fetch to the corresponding XHR object][servo-xhr-enum].
+[Another such enum][servo-constellation-enum] is used for communication between the constellation and the compositor/script.
+
+
+This gives us a great degree of type safety; I can send messages with different data within them,
+however I can only send messages that the other end will know how to handle since they must all be of the type of the message enum.
+
+In Go there's no obvious way to get this. The closest thing is the type called `interface {}`
+which is similar to `Box<Any>` in Rust or `Object` in Java. This is a pointer to any type,
+with the ability to match on its type. As a Rustacean I felt incredibly dirty using this, since I expected that
+there would be an additional vtable overhead &mdash; in Rust `Box<Any>` or `Box<Trait>` are generally avoided in favor of
+generics (for compile time matching) or enums (for runtime matching). Besides, `interface{}` can be fed any type,
+so I can always accidentally send a message of the wrong type through a channel and
+it'll end up crashing the other end at runtime since it hit a `default:` case or something.
+
+
+Of course, I could implement a custom interface `MyMessage` on the various types,
+but this will behave exactly like `interface{}` (implemented on all types) unless I add a dummy method to it, which seems hackish.
+This brings me to my next point:
+
+
+[servo-xhr-enum]: https://github.com/servo/servo/blob/d5dd1d658e5d79701fb9d028479a0fcb26a033fa/components/script/dom/xmlhttprequest.rs#L100
+[servo-constellation-enum]: https://github.com/servo/servo/blob/d5dd1d658e5d79701fb9d028479a0fcb26a033fa/components/msg/constellation_msg.rs#L198
+
+### Smart interfaces
+
+This is something many would consider a feature in Go, but from the point of view of a Rustacean, I'm rather annoyed by this.
```

Comment:
annoyed has a fairly negative connotation. I might rephrase it to something like:

> coming from Rust, Smart interfaces felt almost too magical, and sometimes tripped me up during refactors.